Mill build resolves #274

Sailsman63 opened the pull request:
Draft for moving toward a more flexible build system. It brings support for:

- Managed dependencies (Mostly, we've got a couple of weird ones that are still checked into the repository for now)
    - I've taken the opportunity to upgrade all of our dependencies, as well. Good news: groovy 4.0.0 no longer needs reflection into *everything* just to launch.
- ~~Simpler integration to testing frameworks (not done yet, but removes the need for the builder to manage JUnit separately. I will need to refactor the test layout for this, but it's something I've been considering anyway)~~ EDIT: as of 03-27, the tests have been re-ordered to work directly from mill
- (IMO) A more streamlined, easier to read build process; which will be easier to evolve as the project does forward

**Why this weird mill thing? Why not Maven or Gradle?**

Short version? I like it better.

- Maven has that terribly verbose XML syntax thing going on
- Gradle has a DSL that looks like it was intended as a code-golf language. You really need to study the entire thing to be able to do anything more advanced than "copy-n-paste"
- Both Maven and Gradle have absolutely horrible beginner documentation. It's a convoluted mess of very specific "Recipe" tutorials and technical references without a decent overview of how the architecture works.

In contrast:

- Mill is just verbose enough that I can figure out what it's doing just by skimming.
- It has a nice, short [beginner-friendly user documentation](https://com-lihaoyi.github.io/mill/) that can be read in a single sitting, and gives a foundation for most of the things that one would like to do with any regularity.
- I was able to put together a functional build for AOI (Which is a bit complex as builds go) in less time than I expect it would have taken to do the same in Gradle. Despite the fact that I bumped my head on some FP/Scala-isms on the way.

**Why the huge number of changed files? What are you changing?**
Most of the changed files are not actually *changes*. Resource files have been moved from `{module}/src/` to `{module}/resources/`, which makes tracking them just a bit easier. We should probably go with such a change even if we ultimately reject mill.

I would like feedback/pushback before going this route. What do you think?

EDIT: If you want to just test the build without digging in much, you can run ,s.`./millw Suite.launch` (`./millw.bat Suite.launch` for windows.)</s> for a full build. **EDIT: Now the comands are `./millw run` or `./millw.bat run`.**

peteihis commented:
I haven't looked at this but, (OOT) when it comes to Cradle it did non seem very reliable to me with the STL-thing. While Ant compiles the entire AoI on a 12 thread engine in a few seconds, the time that Gradle spent on the STLT was varying _a lot_ sometimes taking up to three times as long as Ant on AoI.

Sailsman63 commented:
Huh... that sounds weird. Gradle does have its "GradleDaemon" thing going.

I haven't been using a stopwatch, but so far the mill builds seem reasonably fast. The first time you run it, or any time that you change the script, dependencies, or clean the build, it does take a little bit longer, but not bad.

makiam commented:
Try to run build from this branch....
Execute millw.bat Suite.launch
I recieve two errors in output

`Suite.localDeploy java.net.URISyntaxException: Illegal character in path at index 90: https://jogamp.org/deployment/archive/rc/v2.4.0-rc-20210111/jar/gluegen-rt-natives-windows 10-amd64.jar `

and 

`Exception in thread "proxyInputStreamThroughPumper" java.lang.RuntimeException: java.io.IOException: GetOverlappedResult() failed for read operation: 233                                                                  at mill.main.client.InputPumper.run(InputPumper.java:51)                                                                                                                                                           at java.base/java.lang.Thread.run(Thread.java:830)                                                                                                                                                         Caused by: java.io.IOException: GetOverlappedResult() failed for read operation: 233  `

makiam commented:
Another concern (probably not about build process but about development process) is very limited support from developemnt tools to this script.
Once I create IDEA project I open it in IDE and setup to run application via mill script. But no more. For example I did not find how to run debug for project.
From other point gradle (which I preferred) has fill fledged support from IDE.

Sailsman63 commented:
I just pushed a fix for the error with the Windows natives. apparently, `os.name` can vary version to version. The second error was probably due to the first.

I'm doing some digging on how to properly do debug on this style of script. I can run debug (with breakpoints!) by setting up a run configuration for a "Jar Application" and adding the VM option `-agentlib:jdwp=transport=dt_socket,server=y,address=8000,suspend=y`. I agree that there should probably be a better way to do this.

makiam commented:
The last exception mentioned above is due unfixed error in build system. See https://github.com/com-lihaoyi/mill/issues/1310

Sailsman63 commented:
A bit of cleanup and I've migrated the tests to the appropriate modules.

RE: Intellij Idea integration:

- The mill system supports BuildServerProtocol, so we can use the actual build system to manage project structure in the IDE. For Idea users, run `mill mill.bsp.BSP/install` once on the command line before trying to import the project. This will get you incremental compilation, your autocomplete will be able to see into dependencies, and you will be able to run the test suite with a single click.
- Unfortunately, Idea's support for BSP is limited. For debugging, you will need to create two run configurations:
    - One "shell script" config. Use the text `mill Suite.localDeploy`
    - One "Jar Application config.
        - Point to `out/Suite/localDeploy.dest/ArtOfIllusion.jar`
        - Add the Shell Script config to the "Run Before" box.
    - The second config can be targeted either for a simple run, or for debugging. Breakpoints, etc. do function in this configuration. Setting this up should be a one-time thing, unless the script gets significantly refactored.

Give it a try, and tell me how it works!

makiam commented:
BTW... How about to layout sources with maven-style layout?

makiam commented:
This is my configuration for build step
![изображение](https://user-images.githubusercontent.com/16633367/160281331-2676693a-8243-429e-bb93-a41ce47a03f0.png)
And once I invoke build I see at the end of output
[info] done compiling
....
[155/155] Suite.localDeploy 
PS D:\Temp\mill\ArtOfIllusion> 

If I run bulld again I see only this
[129/155] Translators.sources 
PS D:\Temp\mill\ArtOfIllusion> 

And this is my config to run application
![изображение](https://user-images.githubusercontent.com/16633367/160281572-af70c3f1-d449-4d98-ad15-2b2baea5b68c.png)

Once I run it I see only this error message
![изображение](https://user-images.githubusercontent.com/16633367/160281609-de0ca079-1101-4ee7-aea7-c773a7d44647.png)

Sailsman63 commented:
Now *that* one is [an IDEA bug](https://youtrack.jetbrains.com/issue/IDEA-277486).

How often do you expect to need the debug connection? I think I've only used it on a couple of issues...

makiam commented:
javadoc task fails for me and shows this exception:
Suite.javadoc java.io.IOException: Cannot run program "javadoc" (in directory "D:\Tmp\AOI\out\Suite\javadoc.dest"): CreateProcess error=206, File Name or its extension is too long
Once I move AOI folder one level upper this error disappears

Second issue I see that javadoc fails to process some class and fails. I create #278 PR to fix this

Sailsman63 commented:
Ahh... another windows quirk. I think I can work around it, but can you run `mill ArtOfIllusion.docjar` to see if you get the same error?

makiam commented:
No error with docJar

Sailsman63 commented:
Try this one. It puts the source file paths in a file instead of trying to pass them as part of the command line.

lefou commented:
As we discussed this PR a bit in https://github.com/com-lihaoyi/mill/discussions/1463 without me having seen any code, I had a quick glance at your buildfile. I made minimal changes to successfully compile and run it on my machine (Linux):

<details><summary>Here are my changes:</summary>

```diff
@@ -2,11 +2,11 @@ import mill._
 import mill.define.Sources
 import mill.modules.Jvm
 import mill.modules.Util.download
 import scalalib._
 
-trait Common extends JavaModule {
+trait Common extends JavaModule { common =>
 
   def ivyDeps = T {
     super.ivyDeps() ++ Agg(
     ivy"gov.nist.math:jama:1.0.3",
     ivy"com.fifesoft:rsyntaxtextarea:3.1.6",
@@ -15,15 +15,20 @@ trait Common extends JavaModule {
   }
 
   def unmanagedClasspath = T {
     super.unmanagedClasspath() ++
     Agg(
-    PathRef(millSourcePath / os.up / "lib" / "Buoy.jar"),
-    PathRef(millSourcePath / os.up / "lib" / "Buoyx.jar"),
-    PathRef(millSourcePath / os.up / "lib" / "QuickTimeWriter.jar")
+    PathRef(T.workspace / "lib" / "Buoy.jar"),
+    PathRef(T.workspace / "lib" / "Buoyx.jar"),
+    PathRef(T.workspace / "lib" / "QuickTimeWriter.jar")
   ) ++ Cache.unmanagedFiles()
   }
+
+  trait Tests extends super.Tests with TestModule.Junit4 {
+    override def unmanagedClasspath = common.unmanagedClasspath()
+    override def runClasspath = super.runClasspath() ++ ArtOfIllusion.upstreamAssemblyClasspath()
+  }
 }
 
 object ArtOfIllusion extends Common {
 
   def mainClass = Some("artofillusion.ArtOfIllusion")
@@ -48,23 +53,19 @@ object ArtOfIllusion extends Common {
     super.manifest()
       .add("Implementation-Version" -> gitVersion.apply())
       .add("Class-Path" -> libJarPaths())
   }
 
-  object test extends Tests with TestModule.Junit4 with Common {
-    def runClasspath = super.runClasspath() ++ ArtOfIllusion.upstreamAssemblyClasspath()
-  }
+  object test extends Tests
 }
 
 trait PluginModule extends Common {
   def moduleDeps = Seq(ArtOfIllusion)
 }
 
 object Renderers extends PluginModule {
-  object test extends Tests with TestModule.Junit4 with Common {
-    def runClasspath = super.runClasspath() ++ ArtOfIllusion.upstreamAssemblyClasspath()
-  }
+  object test extends Tests
 }
 
 object Filters extends PluginModule
 
 object OSSpecific extends PluginModule
```

</details>

<details><summary>Here is the run output:</summary>

```
10:35:47 lefou:/tmp/ArtOfIllusion mill-build(+18/-10)* 2s 130 ± mill clean
No mill version specified.
You should provide a version via '.mill-version' file or --mill-version option.
Using mill version 0.10.2
[1/1] clean 
10:35:50 θ62° lefou:/tmp/ArtOfIllusion mill-build(+18/-10)* ± mill __.test
No mill version specified.
You should provide a version via '.mill-version' file or --mill-version option.
Using mill version 0.10.2
[27/109] ArtOfIllusion.compile 
[info] compiling 389 Java sources to /tmp/ArtOfIllusion/out/ArtOfIllusion/compile.dest/classes ...
[warn] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/Scene.java:825:1: deprecated item is not annotated with @Deprecated
[warn] public void setSelection(int which)
[warn]   {
[warn]     clearSelection();
[warn]     addToSelection(which);
[warn]     updateSelectionInfo();
[warn]   }
[warn] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/Scene.java:837:1: deprecated item is not annotated with @Deprecated
[warn] public void setSelection(int which[])
[warn]   {
[warn]     clearSelection();
[warn]     for (int index : which) {
[warn]       ObjectInfo info = objects.elementAt(index);
[warn]       if (!info.selected)
[warn]         selection.addElement(index);
[warn]       info.selected = true;
[warn]     }
[warn]     updateSelectionInfo();
[warn]   }
[warn] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/Scene.java:854:1: deprecated item is not annotated with @Deprecated
[warn] public void addToSelection(int which)
[warn]   {
[warn]     ObjectInfo info = objects.elementAt(which);
[warn]     if (!info.selected)
[warn]       selection.addElement(which);
[warn]     info.selected = true;
[warn]     updateSelectionInfo();
[warn]   }
[warn] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/Scene.java:868:1: deprecated item is not annotated with @Deprecated
[warn] public void clearSelection()
[warn]   {
[warn]     if (selection.isEmpty()) return;
[warn]     selection.clear();
[warn]     objects.forEach(item -> { item.selected = false; });
[warn] 
[warn]     updateSelectionInfo();
[warn]   }
[warn] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/Scene.java:882:1: deprecated item is not annotated with @Deprecated
[warn] public void removeFromSelection(int which)
[warn]   {
[warn]     ObjectInfo info = objects.elementAt(which);
[warn]     selection.removeElement(which);
[warn]     info.selected = false;
[warn]     updateSelectionInfo();
[warn]   }
[warn] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/Scene.java:1100:1: deprecated item is not annotated with @Deprecated
[warn] public int [] getSelection()
[warn]   {
[warn]     int sel[] = new int [selection.size()];
[warn] 
[warn]     for (int i = 0; i < sel.length; i++)
[warn]       sel[i] = selection.elementAt(i);
[warn]     return sel;
[warn]   }
[warn] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/Scene.java:1115:1: deprecated item is not annotated with @Deprecated
[warn] public int [] getSelectionWithChildren()
[warn]   {
[warn]     int count = 0;
[warn]     for (int i = objects.size()-1; i >= 0; i--)
[warn]       {
[warn]         ObjectInfo info = objects.elementAt(i);
[warn]         if (info.selected || info.parentSelected)
[warn]           count++;
[warn]       }
[warn]     int sel[] = new int [count];
[warn]     count = 0;
[warn]     for (int i = objects.size()-1; i >= 0; i--)
[warn]       {
[warn]         ObjectInfo info = objects.elementAt(i);
[warn]         if (info.selected || info.parentSelected)
[warn]           sel[count++] = i;
[warn]       }
[warn]     return sel;
[warn]   }
[warn] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/Camera.java:180:1: deprecated item is not annotated with @Deprecated
[warn] public void setSize(int newHres, int newVres)
[warn]   {
[warn]     hres = newHres;
[warn]     vres = newVres;
[warn]     if (perspective)
[warn]       viewToScreen = Mat4.perspective(viewDist);
[warn]     else
[warn]       viewToScreen = Mat4.identity();
[warn]     viewToScreen = Mat4.scale(-scale, -scale, scale).times(viewToScreen);
[warn]     viewToScreen = Mat4.translation((double) hres/2.0, (double) vres/2.0, 0.0).times(viewToScreen);
[warn]     worldToScreen = viewToScreen.times(worldToView);
[warn]     objectToScreen = worldToScreen.times(objectToWorld);
[warn]   }
[warn] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/image/ImageMap.java:38:1: deprecated item is not annotated with @Deprecated
[warn] public static final int PREVIEW_WIDTH = 50;
[warn] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/image/ImageMap.java:40:1: deprecated item is not annotated with @Deprecated
[warn] public static final int PREVIEW_HEIGHT = 50;
[warn] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/MaterialMappingDialog.java:83:1: non-varargs call of varargs method with inexact argument type for last parameter;
[warn]   cast to java.lang.Object for a varargs call
[warn]   cast to java.lang.Object[] for a non-varargs call and to suppress this warning
[warn] null
[warn] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/ModellingApp.java:28:1: deprecated item is not annotated with @Deprecated
[warn] public class ModellingApp
[warn] {
[warn]   public static final String MAJOR_VERSION = ArtOfIllusion.getMajorVersion();
[warn]   public static final String VERSION = ArtOfIllusion.getVersion();
[warn]   public static final double DIST_TO_SCREEN = Camera.DEFAULT_DISTANCE_TO_SCREEN;
[warn]   public static String currentDirectory;
[warn] 
[warn]   public static void main(String args[])
[warn]   {
[warn]     System.out.println("ModellingApp is deprecated.  Use artofillusion.ArtOfIllusion.main() instead.");
[warn]     ArtOfIllusion.main(args);
[warn]   }
[warn] 
[warn]   /** Get the complete version number of Art of Illusion. */
[warn] 
[warn]   public static String getVersion()
[warn]   {
[warn]     return ArtOfIllusion.getVersion();
[warn]   }
[warn] 
[warn]   /** Get the major part of the version number of Art of Illusion. */
[warn] 
[warn]   public static String getMajorVersion()
[warn]   {
[warn]     return ArtOfIllusion.getMajorVersion();
[warn]   }
[warn] 
[warn]   /** Get the application preferences object. */
[warn]   
[warn]   public static ApplicationPreferences getPreferences()
[warn]   {
[warn]     return ArtOfIllusion.getPreferences();
[warn]   }
[warn] 
[warn]   /** Get a list of all available Renderers. */
[warn] 
[warn]   public static Renderer[] getRenderers()
[warn]   {
[warn]     return (Renderer[]) PluginRegistry.getPlugins(Renderer.class).toArray(new Renderer[0]);
[warn]   }
[warn]   
[warn]   /** Get a list of all installed Plugins. */
[warn] 
[warn]   public static Plugin[] getPlugins()
[warn]   {
[warn]     return (Plugin[]) PluginRegistry.getPlugins(Plugin.class).toArray(new Plugin[0]);
[warn]   }
[warn] 
[warn]   /** Get a list of all available Translators. */
[warn] 
[warn]   public static Translator[] getTranslators()
[warn]   {
[warn]     return (Translator[]) PluginRegistry.getPlugins(Translator.class).toArray(new Translator[0]);
[warn]   }
[warn]   
[warn]   /** Get a list of all available ModellingTools. */
[warn]   
[warn]   public static ModellingTool[] getModellingTools()
[warn]   {
[warn]     return (ModellingTool[]) PluginRegistry.getPlugins(ModellingTool.class).toArray(new ModellingTool[0]);
[warn]   }
[warn]   
[warn]   /** Get a list of all available Texture classes. */
[warn] 
[warn]   public static Class[] getTextureTypes()
[warn]   {
[warn]     List instances = PluginRegistry.getPlugins(Texture.class);
[warn]     Class classes[] = new Class[instances.size()];
[warn]     for (int i = 0; i < classes.length; i++)
[warn]       classes[i] = instances.get(i).getClass();
[warn]     return classes;
[warn]   }
[warn]   
[warn]   /** Get a list of all available Material classes. */
[warn] 
[warn]   public static Class[] getMaterialTypes()
[warn]   {
[warn]     List instances = PluginRegistry.getPlugins(Material.class);
[warn]     Class classes[] = new Class[instances.size()];
[warn]     for (int i = 0; i < classes.length; i++)
[warn]       classes[i] = instances.get(i).getClass();
[warn]     return classes;
[warn]   }
[warn]   
[warn]   /** Get a list of all available TextureMapping classes. */
[warn] 
[warn]   public static Class[] getTextureMappings()
[warn]   {
[warn]     List instances = PluginRegistry.getPlugins(TextureMapping.class);
[warn]     Class classes[] = new Class[instances.size()];
[warn]     for (int i = 0; i < classes.length; i++)
[warn]       classes[i] = instances.get(i).getClass();
[warn]     return classes;
[warn]   }
[warn]   
[warn]   /** Get a list of all available MaterialMapping classes. */
[warn] 
[warn]   public static Class[] getMaterialMappings()
[warn]   {
[warn]     List instances = PluginRegistry.getPlugins(MaterialMapping.class);
[warn]     Class classes[] = new Class[instances.size()];
[warn]     for (int i = 0; i < classes.length; i++)
[warn]       classes[i] = instances.get(i).getClass();
[warn]     return classes;
[warn]   }
[warn]   
[warn]   /** Get a list of all available ImageFilter classes. */
[warn] 
[warn]   public static Class[] getImageFilters()
[warn]   {
[warn]     List instances = PluginRegistry.getPlugins(ImageFilter.class);
[warn]     Class classes[] = new Class[instances.size()];
[warn]     for (int i = 0; i < classes.length; i++)
[warn]       classes[i] = instances.get(i).getClass();
[warn]     return classes;
[warn]   }
[warn] 
[warn]   /** Get a list of all plugin-defined procedural Module classes. */
[warn] 
[warn]   public static Class[] getModules()
[warn]   {
[warn]     List instances = PluginRegistry.getPlugins(artofillusion.procedural.Module.class);
[warn]     Class classes[] = new Class[instances.size()];
[warn]     for (int i = 0; i < classes.length; i++)
[warn]       classes[i] = instances.get(i).getClass();
[warn]     return classes;
[warn]   }
[warn]   
[warn]   /** Add a new Renderer the list of available ones. */
[warn]   
[warn]   public static void registerRenderer(Renderer o)
[warn]   {
[warn]     PluginRegistry.registerPlugin(o);
[warn]   }
[warn]   
[warn]   /** Add a new Translator the list of available ones. */
[warn]   
[warn]   public static void registerTranslator(Translator o)
[warn]   {
[warn]     PluginRegistry.registerPlugin(o);
[warn]   }
[warn]   
[warn]   /** Add a new ModellingTool the list of available ones. */
[warn]   
[warn]   public static void registerModellingTool(ModellingTool o)
[warn]   {
[warn]     PluginRegistry.registerPlugin(o);
[warn]   }
[warn]   
[warn]   /** Add a new Texture the list of available ones. */
[warn]   
[warn]   public static void registerTexture(Texture o)
[warn]   {
[warn]     PluginRegistry.registerPlugin(o);
[warn]   }
[warn]   
[warn]   /** Add a new Material the list of available ones. */
[warn]   
[warn]   public static void registerMaterial(Material o)
[warn]   {
[warn]     PluginRegistry.registerPlugin(o);
[warn]   }
[warn]   
[warn]   /** Add a new TextureMapping the list of available ones. */
[warn]   
[warn]   public static void registerTextureMapping(TextureMapping o)
[warn]   {
[warn]     PluginRegistry.registerPlugin(o);
[warn]   }
[warn]   
[warn]   /** Add a new MaterialMapping the list of available ones. */
[warn]   
[warn]   public static void registerMaterialMapping(MaterialMapping o)
[warn]   {
[warn]     PluginRegistry.registerPlugin(o);
[warn]   }
[warn]   
[warn]   /** Add a new Plugin the list of available ones. */
[warn]   
[warn]   public static void registerPlugin(Plugin o)
[warn]   {
[warn]     PluginRegistry.registerPlugin(o);
[warn]   }
[warn]   
[warn]   /** Add a new ImageFilter to the list of available ones. */
[warn]   
[warn]   public static void registerImageFilter(ImageFilter o)
[warn]   {
[warn]     PluginRegistry.registerPlugin(o);
[warn]   }
[warn]   
[warn]   /** Add a new Module to the list of available ones. */
[warn]   
[warn]   public static void registerModule(artofillusion.procedural.Module o)
[warn]   {
[warn]     PluginRegistry.registerPlugin(o);
[warn]   }
[warn]   
[warn]   /** Create a new Scene, and display it in a window. */
[warn]   
[warn]   public static void newWindow()
[warn]   {
[warn]     ArtOfIllusion.newWindow();
[warn]   }
[warn]   
[warn]   /** Create a new window for editing the specified scene. */
[warn]   
[warn]   public static void newWindow(Scene theScene)
[warn]   {
[warn]     ArtOfIllusion.newWindow(theScene);
[warn]   }
[warn]   
[warn]   /** Add a window to the list of open windows. */
[warn]   
[warn]   public static void addWindow(EditingWindow win)
[warn]   {
[warn]     ArtOfIllusion.addWindow(win);
[warn]   }
[warn]   
[warn]   /** Close a window. */
[warn]   
[warn]   public static void closeWindow(EditingWindow win)
[warn]   {
[warn]     ArtOfIllusion.closeWindow(win);
[warn]   }
[warn]   
[warn]   /** Get a list of all open windows. */
[warn]   
[warn]   public static EditingWindow[] getWindows()
[warn]   {
[warn]     return ArtOfIllusion.getWindows();
[warn]   }
[warn]   
[warn]   /** Quit Art of Illusion. */
[warn]   
[warn]   public static void quit()
[warn]   {
[warn]     ArtOfIllusion.quit();
[warn]   }
[warn]   
[warn]   /** Get a class specified by name.  This checks both the system classes, and all plugins.
[warn]       It also accounts for classes which changed packages in version 1.3. */
[warn]   
[warn]   public static Class getClass(String name) throws ClassNotFoundException
[warn]   {
[warn]     return ArtOfIllusion.getClass(name);
[warn]   }
[warn]   
[warn]   /** This is a utility routine which loads a file from disk. */
[warn]   
[warn]   public static String loadFile(File f) throws IOException
[warn]   {
[warn]     return ArtOfIllusion.loadFile(f);
[warn]   }
[warn]   
[warn]   /** Save a scene to a file.  This method returns true if the scene is successfully saved,
[warn]       false if an error occurs. */
[warn] 
[warn]   public static boolean saveScene(Scene sc, LayoutWindow fr)
[warn]   {
[warn]     return ArtOfIllusion.saveScene(sc, fr);
[warn]   }
[warn]   
[warn]   /** Prompt the user to select a scene file, then open a new window containing it.  The BFrame is used for
[warn]       displaying dialogs. */
[warn]   
[warn]   public static void openScene(BFrame fr)
[warn]   {
[warn]     ArtOfIllusion.openScene(fr);
[warn]   }
[warn]     
[warn]   /** Load a scene from a file, and open a new window containing it.  The BFrame is used for
[warn]       displaying dialogs. */
[warn]   
[warn]   public static void openScene(File f, BFrame fr)
[warn]   {
[warn]     ArtOfIllusion.openScene(f, fr);
[warn]   }
[warn]   
[warn]   /** Copy a list of objects to the clipboard, so they can be pasted into either the same scene or a
[warn]       different one. */
[warn]   
[warn]   public static void copyToClipboard(ObjectInfo obj[], Scene scene)
[warn]   {
[warn]     ArtOfIllusion.copyToClipboard(obj, scene);
[warn]   }
[warn]   
[warn]   /** Paste the contents of the clipboard into a window. */
[warn]   
[warn]   public static void pasteClipboard(LayoutWindow win)
[warn]   {
[warn]     ArtOfIllusion.pasteClipboard(win);
[warn]   }
[warn]   
[warn]   /** Get the number of objects on the clipboard. */
[warn]   
[warn]   public static int getClipboardSize()
[warn]   {
[warn]     return ArtOfIllusion.getClipboardSize();
[warn]   }
[warn] }
[warn] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/ObjectPropertiesPanel.java:208:1: non-varargs call of varargs method with inexact argument type for last parameter;
[warn]   cast to java.lang.Object for a varargs call
[warn]   cast to java.lang.Object[] for a non-varargs call and to suppress this warning
[warn] null
[warn] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/ObjectPropertiesPanel.java:255:1: non-varargs call of varargs method with inexact argument type for last parameter;
[warn]   cast to java.lang.Object for a varargs call
[warn]   cast to java.lang.Object[] for a non-varargs call and to suppress this warning
[warn] null
[warn] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/texture/LayeredMapping.java:199:1: deprecated item is not annotated with @Deprecated
[warn] public void addLayer(Texture tex)
[warn]   {
[warn]     Texture newtexture[] = new Texture [texture.length+1];
[warn]     TextureMapping newmapping[] = new TextureMapping [texture.length+1];
[warn]     int newblendMode[] = new int [texture.length+1];
[warn]     int newFractParamID[] = new int [texture.length+1];
[warn]     int i;
[warn] 
[warn]     newtexture[0] = tex;
[warn]     newmapping[0] = tex.getDefaultMapping(theObject);
[warn]     newblendMode[0] = BLEND;
[warn]     newFractParamID[0] = TextureParameter.getUniqueID();
[warn]     for (i = 0; i < texture.length; i++)
[warn]       {
[warn]         newtexture[i+1] = texture[i];
[warn]         newmapping[i+1] = mapping[i];
[warn]         newblendMode[i+1] = blendMode[i];
[warn]         newFractParamID[i+1] = fractParamID[i];
[warn]       }
[warn]     texture = newtexture;
[warn]     mapping = newmapping;
[warn]     blendMode = newblendMode;
[warn]     fractParamID = newFractParamID;
[warn]   }
[warn] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/ui/ThemeManager.java:559:1: deprecated item is not annotated with @Deprecated
[warn] public static ToolButton getToolButton(Object owner, String iconName, String selectedIconName)
[warn]     {
[warn]         System.out.println("**Deprecated method called: ThemeManager.getToolButton(Object, String, String)");
[warn] 
[warn]         Exception e = new Exception();
[warn]         StackTraceElement[] trace = e.getStackTrace();
[warn] 
[warn]         if (trace.length > 1) {
[warn]             StackTraceElement frame = trace[1];
[warn]             String name = frame.getClassName();
[warn]             int cut = name.lastIndexOf('.');
[warn] 
[warn]             System.out.print("\tcalled from ");
[warn]             if (frame.getFileName() != null) {
[warn]                 System.out.print(frame.getFileName());
[warn]                 System.out.print(':');
[warn]                 System.out.print(String.valueOf(frame.getLineNumber()));
[warn]             }
[warn]             else {
[warn]                 System.out.print(cut > 0 ? name.substring(cut+1) : name);
[warn]                 System.out.print('.');
[warn]                 System.out.print(frame.getMethodName());
[warn]                 System.out.print("() (unknown source)");
[warn]             }
[warn] 
[warn]             System.out.println();
[warn]         }
[warn] 
[warn]         return getToolButton(owner, iconName);
[warn]     }
[warn] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/RenderingDialog.java:294:1: non-varargs call of varargs method with inexact argument type for last parameter;
[warn]   cast to java.lang.Object for a varargs call
[warn]   cast to java.lang.Object[] for a non-varargs call and to suppress this warning
[warn] new String [] {current, total, Integer.toString(currentSubimage+1), status}
[warn] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/RenderingDialog.java:296:1: non-varargs call of varargs method with inexact argument type for last parameter;
[warn]   cast to java.lang.Object for a varargs call
[warn]   cast to java.lang.Object[] for a non-varargs call and to suppress this warning
[warn] new String [] {current, total, status}
[warn] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/TextureMappingDialog.java:96:1: non-varargs call of varargs method with inexact argument type for last parameter;
[warn]   cast to java.lang.Object for a varargs call
[warn]   cast to java.lang.Object[] for a non-varargs call and to suppress this warning
[warn] null
[info] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/LayoutWindow.java: Some input files use or override a deprecated API.
[info] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/LayoutWindow.java: Recompile with -Xlint:deprecation for details.
[info] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/Scene.java: Some input files use unchecked or unsafe operations.
[info] /tmp/ArtOfIllusion/ArtOfIllusion/src/artofillusion/Scene.java: Recompile with -Xlint:unchecked for details.
[info] done compiling
[44/109] ArtOfIllusion.test.compile 
[info] compiling 16 Java sources to /tmp/ArtOfIllusion/out/ArtOfIllusion/test/compile.dest/classes ...
[info] /tmp/ArtOfIllusion/ArtOfIllusion/test/src/artofillusion/procedural/CompareModuleTest.java: Some input files use or override a deprecated API.
[info] /tmp/ArtOfIllusion/ArtOfIllusion/test/src/artofillusion/procedural/CompareModuleTest.java: Recompile with -Xlint:deprecation for details.
[info] done compiling
[58/109] ArtOfIllusion.test.test 
Test run started
Test artofillusion.texture.LayeredTextureTest.testHasComponent started
Test artofillusion.texture.LayeredTextureTest.testHasComponent finished, took 0.007 sec
Test artofillusion.texture.LayeredTextureTest.testParameters started
Test artofillusion.texture.LayeredTextureTest.testParameters finished, took 0.384 sec
Test run finished: 0 failed, 0 ignored, 2 total, 0.399s
Test run started
Test artofillusion.object.ObjectInfoTest.testDuplicateWithNewGeometry started
Test artofillusion.object.ObjectInfoTest.testDuplicateWithNewGeometry finished, took 0.001 sec
Test artofillusion.object.ObjectInfoTest.testGetDistortedObjectFromWrapperNoDistortion started
Test artofillusion.object.ObjectInfoTest.testGetDistortedObjectFromWrapperNoDistortion finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testRemoveTrackByPositionLast started
Test artofillusion.object.ObjectInfoTest.testRemoveTrackByPositionLast finished, took 0.003 sec
Test artofillusion.object.ObjectInfoTest.testAddDistortionToNull started
Test artofillusion.object.ObjectInfoTest.testAddDistortionToNull finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testCopyInfoWithTracksOverExisted started
Test artofillusion.object.ObjectInfoTest.testCopyInfoWithTracksOverExisted finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testCopyInfoWithNullDistortionOverExisted started
Test artofillusion.object.ObjectInfoTest.testCopyInfoWithNullDistortionOverExisted finished, took 0.001 sec
Test artofillusion.object.ObjectInfoTest.testAddChildFirst started
Test artofillusion.object.ObjectInfoTest.testAddChildFirst finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testRemoveTrackByTrack started
Test artofillusion.object.ObjectInfoTest.testRemoveTrackByTrack finished, took 0.001 sec
Test artofillusion.object.ObjectInfoTest.testGetMeshSkeleton started
Test artofillusion.object.ObjectInfoTest.testGetMeshSkeleton finished, took 0.006 sec
Test artofillusion.object.ObjectInfoTest.testRemoveTrackByPositionFirst started
Test artofillusion.object.ObjectInfoTest.testRemoveTrackByPositionFirst finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testGetRenderingMeshWithUnsetTexture started
Test artofillusion.object.ObjectInfoTest.testGetRenderingMeshWithUnsetTexture finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testGetCubeSkeleton started
Test artofillusion.object.ObjectInfoTest.testGetCubeSkeleton finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testAddTrack started
Test artofillusion.object.ObjectInfoTest.testAddTrack finished, took 0.001 sec
Test artofillusion.object.ObjectInfoTest.testRemoveChildByMissedChild started
Test artofillusion.object.ObjectInfoTest.testRemoveChildByMissedChild finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testAddTrackToGivenErrorPos started
Test artofillusion.object.ObjectInfoTest.testAddTrackToGivenErrorPos finished, took 0.001 sec
Test artofillusion.object.ObjectInfoTest.setObjectInfoMaterial started
Test artofillusion.object.ObjectInfoTest.setObjectInfoMaterial finished, took 0.001 sec
Test artofillusion.object.ObjectInfoTest.testDuplicateWithDistortion started
Test artofillusion.object.ObjectInfoTest.testDuplicateWithDistortion finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testAddChildTwo started
Test artofillusion.object.ObjectInfoTest.testAddChildTwo finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testDuplicateWithTracks started
Test artofillusion.object.ObjectInfoTest.testDuplicateWithTracks finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testSetTextureWithTextureTracks started
Test artofillusion.object.ObjectInfoTest.testSetTextureWithTextureTracks finished, took 0.001 sec
Test artofillusion.object.ObjectInfoTest.testCopyInfo started
Test artofillusion.object.ObjectInfoTest.testCopyInfo finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testAddDistortionToExisted started
Test artofillusion.object.ObjectInfoTest.testAddDistortionToExisted finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testDuplicateWithNewGeometryAndTracks started
Test artofillusion.object.ObjectInfoTest.testDuplicateWithNewGeometryAndTracks finished, took 0.001 sec
Test artofillusion.object.ObjectInfoTest.testCopyInfoWithDistortionOverEmpty started
Test artofillusion.object.ObjectInfoTest.testCopyInfoWithDistortionOverEmpty finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testSetDistortion started
Test artofillusion.object.ObjectInfoTest.testSetDistortion finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testClearNotNullDistortion started
Test artofillusion.object.ObjectInfoTest.testClearNotNullDistortion finished, took 0.001 sec
Test artofillusion.object.ObjectInfoTest.testAddTrackToGivenPosInExistList started
Test artofillusion.object.ObjectInfoTest.testAddTrackToGivenPosInExistList finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testGetRenderingMesh started
Test artofillusion.object.ObjectInfoTest.testGetRenderingMesh finished, took 0.001 sec
Test artofillusion.object.ObjectInfoTest.testClearNullDistortion started
Test artofillusion.object.ObjectInfoTest.testClearNullDistortion finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testSetObjectInfoParent started
Test artofillusion.object.ObjectInfoTest.testSetObjectInfoParent finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testCreateNewObjectInfo started
Test artofillusion.object.ObjectInfoTest.testCreateNewObjectInfo finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testDuplicate started
Test artofillusion.object.ObjectInfoTest.testDuplicate finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testDuplicateAll started
Test artofillusion.object.ObjectInfoTest.testDuplicateAll finished, took 0.001 sec
Test artofillusion.object.ObjectInfoTest.testRemoveChildByPositionOne started
Test artofillusion.object.ObjectInfoTest.testRemoveChildByPositionOne finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testRemoveChildByPositionTwo started
Test artofillusion.object.ObjectInfoTest.testRemoveChildByPositionTwo finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testRemoveChildByChild started
Test artofillusion.object.ObjectInfoTest.testRemoveChildByChild finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testRemoveChildByNullChild started
Test artofillusion.object.ObjectInfoTest.testRemoveChildByNullChild finished, took 0.0 sec
Test artofillusion.object.ObjectInfoTest.testGetDistortedObjectNoDistortion started
Test artofillusion.object.ObjectInfoTest.testGetDistortedObjectNoDistortion finished, took 0.0 sec
Test run finished: 0 failed, 0 ignored, 38 total, 0.03s
Test run started
Test artofillusion.object.ImplicitSphereTest.gradient_estimate_within_delta started
Test artofillusion.object.ImplicitSphereTest.gradient_estimate_within_delta finished, took 0.058 sec
Test artofillusion.object.ImplicitSphereTest.point_outside_influence_radius_is_0 started
Test artofillusion.object.ImplicitSphereTest.point_outside_influence_radius_is_0 finished, took 0.01 sec
Test artofillusion.object.ImplicitSphereTest.point_inside_radius_is_greater_than_1 started
Test artofillusion.object.ImplicitSphereTest.point_inside_radius_is_greater_than_1 finished, took 0.003 sec
Test artofillusion.object.ImplicitSphereTest.point_between_radius_and_influence_is_between_0_and_1 started
Test artofillusion.object.ImplicitSphereTest.point_between_radius_and_influence_is_between_0_and_1 finished, took 0.007 sec
Test artofillusion.object.ImplicitSphereTest.gradient_estimate_at_influence_edge started
Test artofillusion.object.ImplicitSphereTest.gradient_estimate_at_influence_edge finished, took 0.002 sec
Test run finished: 0 failed, 0 ignored, 5 total, 0.085s
Test run started
Test artofillusion.SceneTest.testGetSceneHasNoCamerasList started
Test artofillusion.SceneTest.testGetSceneHasNoCamerasList finished, took 0.028 sec
Test artofillusion.SceneTest.testGetSceneHasSingleCameraOnly started
Test artofillusion.SceneTest.testGetSceneHasSingleCameraOnly finished, took 0.003 sec
Test artofillusion.SceneTest.testGetSceneHasCamerasOnly started
Test artofillusion.SceneTest.testGetSceneHasCamerasOnly finished, took 0.0 sec
Test artofillusion.SceneTest.testSceneHasCameraAndOther started
Test artofillusion.SceneTest.testSceneHasCameraAndOther finished, took 0.003 sec
Test run finished: 0 failed, 0 ignored, 4 total, 0.036s
Test run started
Test artofillusion.math.SimplexNoiseTest.testRange started
Test artofillusion.math.SimplexNoiseTest.testRange finished, took 0.017 sec
Test artofillusion.math.SimplexNoiseTest.testGradient2D started
Test artofillusion.math.SimplexNoiseTest.testGradient2D finished, took 0.005 sec
Test artofillusion.math.SimplexNoiseTest.testGradient3D started
Test artofillusion.math.SimplexNoiseTest.testGradient3D finished, took 0.009 sec
Test artofillusion.math.SimplexNoiseTest.testVector started
Test artofillusion.math.SimplexNoiseTest.testVector finished, took 0.005 sec
Test run finished: 0 failed, 0 ignored, 4 total, 0.039s
Test run started
Test artofillusion.math.CoordinateSystemEqualityTest.testCustomCSAreEquals started
Test artofillusion.math.CoordinateSystemEqualityTest.testCustomCSAreEquals finished, took 0.001 sec
Test artofillusion.math.CoordinateSystemEqualityTest.testCSNotEqualsByUpDir started
Test artofillusion.math.CoordinateSystemEqualityTest.testCSNotEqualsByUpDir finished, took 0.0 sec
Test artofillusion.math.CoordinateSystemEqualityTest.testCSNotEqualsByOrig started
Test artofillusion.math.CoordinateSystemEqualityTest.testCSNotEqualsByOrig finished, took 0.0 sec
Test artofillusion.math.CoordinateSystemEqualityTest.testCSNotEqualsByZDir started
Test artofillusion.math.CoordinateSystemEqualityTest.testCSNotEqualsByZDir finished, took 0.0 sec
Test artofillusion.math.CoordinateSystemEqualityTest.testCSAreEquals started
Test artofillusion.math.CoordinateSystemEqualityTest.testCSAreEquals finished, took 0.0 sec
Test artofillusion.math.CoordinateSystemEqualityTest.testCustomCSAreEquals2 started
Test artofillusion.math.CoordinateSystemEqualityTest.testCustomCSAreEquals2 finished, took 0.001 sec
Test run finished: 0 failed, 0 ignored, 6 total, 0.005s
Test run started
Test artofillusion.math.RGBColorTest.testERGB started
Test artofillusion.math.RGBColorTest.testERGB finished, took 0.003 sec
Test run finished: 0 failed, 0 ignored, 1 total, 0.004s
Test run started
Test artofillusion.animation.JointEqualityTest.testJointNotEqualsByTwistNotFixed started
Test artofillusion.animation.JointEqualityTest.testJointNotEqualsByTwistNotFixed finished, took 0.001 sec
Test artofillusion.animation.JointEqualityTest.testJointNotEqualsByAngleOne started
Test artofillusion.animation.JointEqualityTest.testJointNotEqualsByAngleOne finished, took 0.0 sec
Test artofillusion.animation.JointEqualityTest.testJointNotEqualsByAngleTwo started
Test artofillusion.animation.JointEqualityTest.testJointNotEqualsByAngleTwo finished, took 0.0 sec
Test artofillusion.animation.JointEqualityTest.testJointNotEqualsByTwistNotComfort started
Test artofillusion.animation.JointEqualityTest.testJointNotEqualsByTwistNotComfort finished, took 0.0 sec
Test artofillusion.animation.JointEqualityTest.testJointNotEqualsByLength started
Test artofillusion.animation.JointEqualityTest.testJointNotEqualsByLength finished, took 0.0 sec
Test artofillusion.animation.JointEqualityTest.testJointNotEqualsByTwist started
Test artofillusion.animation.JointEqualityTest.testJointNotEqualsByTwist finished, took 0.0 sec
Test artofillusion.animation.JointEqualityTest.testJointNotEqualsByName started
Test artofillusion.animation.JointEqualityTest.testJointNotEqualsByName finished, took 0.0 sec
Test artofillusion.animation.JointEqualityTest.testJointEqualsByName started
Test artofillusion.animation.JointEqualityTest.testJointEqualsByName finished, took 0.0 sec
Test artofillusion.animation.JointEqualityTest.testJointNotEqualsByTwistNotLoop started
Test artofillusion.animation.JointEqualityTest.testJointNotEqualsByTwistNotLoop finished, took 0.0 sec
Test run finished: 0 failed, 0 ignored, 9 total, 0.003s
Test run started
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreNotEqualsByLoop started
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreNotEqualsByLoop finished, took 0.0 sec
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreNotEqualsByMax started
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreNotEqualsByMax finished, took 0.0 sec
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreNotEqualsByMin started
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreNotEqualsByMin finished, took 0.0 sec
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreNotEqualsByPos started
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreNotEqualsByPos finished, took 0.0 sec
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreNotEqualsByStiffness started
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreNotEqualsByStiffness finished, took 0.0 sec
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreNotEqualsByMinComfort started
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreNotEqualsByMinComfort finished, took 0.0 sec
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreNotEqualsByComfort started
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreNotEqualsByComfort finished, took 0.0 sec
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreEquals started
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreEquals finished, took 0.001 sec
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreNotEqualsByFixed started
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreNotEqualsByFixed finished, took 0.0 sec
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreNotEqualsByMaxComfort started
Test artofillusion.animation.JointDOFEqualityTest.testDOFAreNotEqualsByMaxComfort finished, took 0.001 sec
Test run finished: 0 failed, 0 ignored, 10 total, 0.003s
Test run started
Test artofillusion.procedural.CompareModuleTest.testCompare started
Test artofillusion.procedural.CompareModuleTest.testCompare finished, took 0.001 sec
Test run finished: 0 failed, 0 ignored, 1 total, 0.002s
Test run started
Test artofillusion.procedural.ImageModuleTest.testHLS started
Test artofillusion.procedural.ImageModuleTest.testHLS finished, took 0.022 sec
Test artofillusion.procedural.ImageModuleTest.testHSV started
Test artofillusion.procedural.ImageModuleTest.testHSV finished, took 0.007 sec
Test artofillusion.procedural.ImageModuleTest.testRGB started
Test artofillusion.procedural.ImageModuleTest.testRGB finished, took 0.006 sec
Test run finished: 0 failed, 0 ignored, 3 total, 0.036s
Test run started
Test artofillusion.procedural.ModuleTest.testModuleDuplicate started
Test artofillusion.procedural.ModuleTest.testModuleDuplicate finished, took 0.001 sec
Test run finished: 0 failed, 0 ignored, 1 total, 0.001s
Test run started
Test artofillusion.procedural.MinAndMaxModulesTest.testMinModuleEquals started
Test artofillusion.procedural.MinAndMaxModulesTest.testMinModuleEquals finished, took 0.001 sec
Test artofillusion.procedural.MinAndMaxModulesTest.testMaxModuleEquals started
Test artofillusion.procedural.MinAndMaxModulesTest.testMaxModuleEquals finished, took 0.0 sec
Test artofillusion.procedural.MinAndMaxModulesTest.testMinModuleSeparated started
Test artofillusion.procedural.MinAndMaxModulesTest.testMinModuleSeparated finished, took 0.0 sec
Test artofillusion.procedural.MinAndMaxModulesTest.testRandoms started
Test artofillusion.procedural.MinAndMaxModulesTest.testRandoms finished, took 0.001 sec
Test artofillusion.procedural.MinAndMaxModulesTest.testMaxModuleSeparated started
Test artofillusion.procedural.MinAndMaxModulesTest.testMaxModuleSeparated finished, took 0.001 sec
Test run finished: 0 failed, 0 ignored, 5 total, 0.003s
Test run started
Test artofillusion.procedural.PowerModuleTest.testValue started
Test artofillusion.procedural.PowerModuleTest.testValue finished, took 0.025 sec
Test artofillusion.procedural.PowerModuleTest.testGradient started
Test artofillusion.procedural.PowerModuleTest.testGradient finished, took 0.001 sec
Test run finished: 0 failed, 0 ignored, 2 total, 0.026s
Test run started
Test artofillusion.util.TestThreadManager.testRun started
Test artofillusion.util.TestThreadManager.testRun finished, took 0.034 sec
Test artofillusion.util.TestThreadManager.testCancel started
Test artofillusion.util.TestThreadManager.testCancel finished, took 0.017 sec
Test run finished: 0 failed, 0 ignored, 2 total, 0.052s
[82/109] Renderers.compile 
[info] compiling 46 Java sources to /tmp/ArtOfIllusion/out/Renderers/compile.dest/classes ...
[info] /tmp/ArtOfIllusion/Renderers/src/artofillusion/raster/Raster.java: Some input files use unchecked or unsafe operations.
[info] /tmp/ArtOfIllusion/Renderers/src/artofillusion/raster/Raster.java: Recompile with -Xlint:unchecked for details.
[info] done compiling
[99/109] Renderers.test.compile 
[info] compiling 1 Java source to /tmp/ArtOfIllusion/out/Renderers/test/compile.dest/classes ...
[info] done compiling
[109/109] Renderers.test.test 
Test run started
Test artofillusion.raytracer.OctreeNodeTest.testFindNode started
Test artofillusion.raytracer.OctreeNodeTest.testFindNode finished, took 0.004 sec
Test artofillusion.raytracer.OctreeNodeTest.testTraceRay started
Test artofillusion.raytracer.OctreeNodeTest.testTraceRay finished, took 0.008 sec
Test run finished: 0 failed, 0 ignored, 2 total, 0.042s

```
</details>

lefou commented on build.sc:
```diff
@@ -0,0 +1,239 @@
+  def unmanagedFiles = T {
+    Agg(
+      downloadFile(
+        "https://github.com/blackears/svgSalamander/releases/download/v1.1.3/svgSalamander-1.1.3.jar"),
+      downloadFile(
+        "https://github.com/beanshell/beanshell/releases/download/2.1.0/bsh-2.1.0.jar"),
+      download("https://maven.jzy3d.org/releases/org/jogamp/gluegen/gluegen-rt/v2.4.0-rc4/gluegen-rt-v2.4.0-rc4.jar",
+        os.rel / "gluegen-rt.jar"),
+      download("https://maven.jzy3d.org/releases/org/jogamp/jogl/jogl-all/v2.4.0-rc4/jogl-all-v2.4.0-rc4.jar",
+        os.rel / "jogl-all.jar")
```
You're manually downloading lots of dependencies from a Maven repository. Why don't you configure the repository via `repositoryTask` and use `ivyDeps` or `runIvyDeps` instead?

Sailsman63 replied:
Any thoughts on how to shade jar names *outside* of the `assembly` target? I've found `Assembly.getShadedClasspath`, but have no idea what to do with a `Generator` - I don't even understand the type signature.

lefou replied:
I have no idea what you mean with "shade jar names".

lefou replied:
Now I understand. The repo isn't really broken, but you need jars without the version number in the filename. How about replacing the jars names after they are resolved? e.g. in `compileClasspath` or `runClasspath`, you could just map over the super target result and copy those that need modification into `T.dest` with the new filename and use these instead.

Sailsman63 replied:
Thanks for the hint. Since this only really matters when *deploying*, I modified the names while copying the files into the deployment assembly. This should only last until we upgrade to a newer version of the library, or replace it altogether.

lefou commented:
It's called a self reference. The concept behind it is called self-type. See https://docs.scala-lang.org/tour/self-types.html

Without a specified type, you can think of it as an "alias" to `this`. Very useful inside of nested classes or traits.

Sailsman63 commented:
Added a thin ant shim for compatibility with IDEs that do not understand mill. Targets for:

- compile
- clean
- run
- run-debug (This one suspends until you connect a debugger. Does not work on NetBeans)
- debug-nb (works on NetBeans' very bespoke system. Please read the comment when first setting up.)

makiam commented:
Netbeans outputs this....
```
ant -f D:\\Tmp\\ArtOfIllusion run
run:
D:\Tmp\ArtOfIllusion\build.xml:22: Execute failed: 
java.io.IOException: Cannot run program "millw.bat" (in directory "D:\Tmp\ArtOfIllusion"): CreateProcess error=2, The specified file cannot be found
```

Sailsman63 commented:
Just improved the shim - apparently, there is some platform-specific behavior to both Netbeans and IDEA regarding ant-calls to outside scripts. Have tested in both IDEs on both Windows 10 and Arch Linux.

lefou commented:
IMHO, platform specific variants are a good candidate for `Cross` modules. You gain stable targets and the option to build all on any platform.

Sailsman63 commented:
That might be worth considering. We've mostly shied away uberjar constructs in the past because:

- Our "platform-specific" is literally limited to the one set of native dependencies - which are optional - and an installer wrapper.
- Most of the structure that is generated in `Suite.stage` is required anyway. We provide a space for user-scripts and third-party plugins. Unless/until we migrate that to something that is configurable and does not require the static directory structure, generating different versions of an "Application Jar" doesn't seem to make sense to me.

makiam commented:
I tnink we may not to separate AOI core application jar of all other libraries and put all jars together.
See this patch for this
[single-jars-location-patch.txt](https://github.com/ArtOfIllusion/ArtOfIllusion/files/8441921/single-jars-location-patch.txt)

Sailsman63 commented:
Certainly we could. What, in your opinion, is the gain?

makiam commented:
IMO it's just simpler to have one location for all jars.

makiam commented:
Update internal QT Writer library to external reference

[replace-qt-writer-with-external-dependency.txt](https://github.com/ArtOfIllusion/ArtOfIllusion/files/8458839/replace-qt-writer-with-external-dependency.txt)

Sailsman63 commented:
I think that we're going to have to hold off on that one. (BTW, you *can* open a PR against this branch, rather than uploading patch files)

Two reasons:

1. This would pull in the entire Monte-Media library, rather than the very tiny subset that makes up the Quicktime Writer.
2. More importantly, I've looked at the updated library on the original author's site. The current version that he offers is licensed under CC BY-SA 3.0, which per [CreativeCommons](https://creativecommons.org/share-your-work/licensing-considerations/compatible-licenses) is *not* compatible with GPL 2 *or* 3. (CC licenses are not designed for software anyway, so it's an awkward license to work with) I'd assume that the original version that we have used to be under a different license.

Sailsman63 commented:
Just pushed out the build and launch commands to project level, so they can be called as `mill launch` or similar. This also removes the `LocalInstall` directory from mill's artifact folder, so plugins and scripts will survive a `mill clean`

Also improved the documentation of this build process.

NB: There are some things in docs/Building.md *not* related to these changes that need to be updated.

Sailsman63 commented:
Just rebased to keep mergeable with the main branch.

- Updated some dependencies that had fallen behind.
- Changed the basic build command to "run" - I learned coming back to this after a hiatus that this would be more intuitive.

makiam commented:
Try to run build
See an exception: 
[31mjava.util.NoSuchElementException: key not found: List(win)
	at scala.collection.immutable.Map$Map4.apply(Map.scala:517)
	at mill.define.Cross.apply(Cross.scala:85)
	at ammonite.$file.build.installLocal(D:\Tmp\Apps\ArtOfIllusion\build.sc:202)
	at ammonite.$file.build.installLocal$(D:\Tmp\Apps\ArtOfIllusion\build.sc:201)
	at ammonite.$file.build$.installLocal(D:\Tmp\Apps\ArtOfIllusion\build.sc:41)
	at ammonite.$file.build.run(D:\Tmp\Apps\ArtOfIllusion\build.sc:211)
	at ammonite.$file.build.run$(D:\Tmp\Apps\ArtOfIllusion\build.sc:210)
	at ammonite.$file.build$.run(D:\Tmp\Apps\ArtOfIllusion\build.sc:41)
	at ammonite.$file.build$.$anonfun$millDiscover$82(D:\Tmp\Apps\ArtOfIllusion\build.sc:54)

Sailsman63 commented:
java version? Are you running with an installed version of mill, or are you using the millw wrapper?

makiam commented:
I run this with wrapper on jdk 18.0.2

Sailsman63 commented:
Got it. I had a typo in the platform-build selector.
